File: code/dart/modular_api/example/modules/time/usecases/current_time.dart
import 'package:modular_api/modular_api.dart';

// ─── Input DTO ────────────────────────────────────────────────────────────────

class CurrentTimeInput extends Input {
  final String tz;

  CurrentTimeInput({this.tz = ''});

  @override
  Map<String, dynamic> toJson() => {'tz': tz};

  @override
  List<SchemaField> get schemaFields => [
        SchemaField.optional(
          SchemaField.string('tz',
              description: 'Timezone offset (e.g. utc-5, utc+3, utc)',
              example: 'utc-5'),
        ),
      ];

  static CurrentTimeInput get example => CurrentTimeInput(tz: 'utc-5');
}

// ─── Output DTO ───────────────────────────────────────────────────────────────

class CurrentTimeOutput extends Output {
  final String datetime;
  final int offset;

  CurrentTimeOutput({required this.datetime, required this.offset});

  @override
  int get statusCode => 200;

  @override
  Map<String, dynamic> toJson() => {'datetime': datetime, 'offset': offset};

  @override
  List<SchemaField> get schemaFields => [
        SchemaField.string('datetime',
            description: 'ISO 8601 datetime at the requested offset',
            example: '2026-03-14T07:00:00'),
        SchemaField.integer('offset',
            description: 'UTC offset in hours', example: -5),
      ];

  static CurrentTimeOutput get example =>
      CurrentTimeOutput(datetime: '2026-03-14T07:00:00', offset: -5);
}

// ─── UseCase ──────────────────────────────────────────────────────────────────

class CurrentTime implements UseCase<CurrentTimeInput, CurrentTimeOutput> {
  @override
  final CurrentTimeInput input;

  @override
  ModularLogger? logger;

  CurrentTime({required this.input});

  static CurrentTime fromJson(Map<String, dynamic> json) {
    return CurrentTime(
        input: CurrentTimeInput(tz: (json['tz'] ?? '').toString()));
  }

  @override
  String? validate() {
    if (input.tz.isEmpty) return null;
    final offset = _parseOffset(input.tz);
    if (offset == null) return 'invalid timezone format, use utc, utc-5, utc+3';
    if (offset < -12 || offset > 14) {
      return 'offset must be between -12 and +14';
    }
    return null;
  }

  @override
  Future<CurrentTimeOutput> execute() async {
    final now = DateTime.now().toUtc();
    final offsetHours = input.tz.isEmpty
        ? DateTime.now().timeZoneOffset.inHours
        : _parseOffset(input.tz)!;
    final adjusted = now.add(Duration(hours: offsetHours));
    final iso = adjusted.toIso8601String().split('.').first;

    logger?.info('Time requested for offset $offsetHours');
    return CurrentTimeOutput(datetime: iso, offset: offsetHours);
  }

  /// Parses "utc-5", "utc+3", "utc" into an integer offset. Returns null on bad format.
  static int? _parseOffset(String tz) {
    final lower = tz.toLowerCase().trim();
    if (lower == 'utc') return 0;
    final match = RegExp(r'^utc([+-]\d{1,2})$').firstMatch(lower);
    if (match == null) return null;
    return int.tryParse(match.group(1)!);
  }
}

File: code/ts/modular_api/example/modules/time/usecases/current_time.ts
import { Input, Output, UseCase, Field, ModularLogger } from '../../../../src/index';

// ─── Input DTO ────────────────────────────────────────────────────────────────

export class CurrentTimeInput extends Input {
  @Field.optional(Field.string({ description: 'Timezone offset (e.g. utc-5, utc+3, utc)', example: 'utc-5' }))
  tz?: string;
}

// ─── Output DTO ───────────────────────────────────────────────────────────────

export class CurrentTimeOutput extends Output {
  @Field.string({ description: 'ISO 8601 datetime at the requested offset', example: '2026-03-14T07:00:00' })
  datetime!: string;

  @Field.integer({ description: 'UTC offset in hours', example: -5 })
  offset!: number;

  get statusCode() {
    return 200;
  }
}

// ─── UseCase ──────────────────────────────────────────────────────────────────

export class CurrentTime implements UseCase<CurrentTimeInput, CurrentTimeOutput> {
  readonly input: CurrentTimeInput;
  logger?: ModularLogger;

  constructor(input: CurrentTimeInput) {
    this.input = input;
  }

  static fromJson(json: Record<string, unknown>): CurrentTime {
    const input = new CurrentTimeInput();
    input.tz = json['tz'] != null ? String(json['tz']) : undefined;
    return new CurrentTime(input);
  }

  validate(): string | null {
    if (!this.input.tz) return null;
    const offset = CurrentTime.parseOffset(this.input.tz);
    if (offset === null) return 'invalid timezone format, use utc, utc-5, utc+3';
    if (offset < -12 || offset > 14) return 'offset must be between -12 and +14';
    return null;
  }

  async execute(): Promise<CurrentTimeOutput> {
    const now = new Date();
    const offsetHours = this.input.tz
      ? CurrentTime.parseOffset(this.input.tz)!
      : -(now.getTimezoneOffset() / 60);
    const adjusted = new Date(now.getTime() + offsetHours * 3600_000);
    const iso = adjusted.toISOString().split('.')[0];

    this.logger?.info(`Time requested for offset ${offsetHours}`);
    const output = new CurrentTimeOutput();
    output.datetime = iso;
    output.offset = offsetHours;
    return output;
  }

  /** Parses "utc-5", "utc+3", "utc" into an integer offset. Returns null on bad format. */
  private static parseOffset(tz: string): number | null {
    const lower = tz.toLowerCase().trim();
    if (lower === 'utc') return 0;
    const match = lower.match(/^utc([+-]\d{1,2})$/);
    if (!match) return null;
    return parseInt(match[1], 10);
  }
}
